radiometric_normalization/pif.py

```diff
@@ -72,3 +81,156 @@ def _filter_zero_alpha_pifs(reference_gimage, candidate_gimage):
+    alpha = numpy.ones(array_shape)
+    alpha[numpy.nonzero(candidate_gimage.alpha == 0)] = 0
+    alpha[numpy.nonzero(reference_gimage.alpha == 0)] = 0
+    alpha_vec = alpha.flatten()
+    valid_pixels_list = numpy.nonzero(alpha_vec != 0)
+
+    PIF_all = numpy.ones(array_shape)
+
+    # Per band analysis
+    for band_no in xrange(len(candidate_gimage.bands)):
+
+        logging.info('PCA Info: Band {}'.format(band_no))
+
+        PIF_band = _PCA_fit_and_filter_single_band(
+            candidate_gimage.bands[band_no].flatten(),
+            reference_gimage.bands[band_no].flatten(),
+            valid_pixels_list, array_shape, lim)
+        PIF_all = numpy.logical_and(PIF_band, PIF_all)
+
+    if logging.getLogger().getEffectiveLevel() <= logging.INFO:
+        for band_no in xrange(len(candidate_gimage.bands)):
+            cand_combo_filtered = candidate_gimage.bands[band_no][
+                numpy.nonzero(PIF_all != 0)].flatten()
+            ref_combo_filtered = reference_gimage.bands[band_no][
+                numpy.nonzero(PIF_all != 0)].flatten()
+
+            logging.info('PCA Info: Filtered combo corrcoef (band {}) ='
+                         ' {}'.format(band_no, numpy.corrcoef(
+                            cand_combo_filtered, ref_combo_filtered)[0, 1]))
+
+    no_valid_pixels = len(numpy.nonzero(PIF_all.flatten() != 0)[0])
+    no_total_pixels = candidate_gimage.bands[0].size
+    valid_percent = 100.0 * no_valid_pixels / no_total_pixels
+    logging.info('PCA Info: Found {} final pifs out of {} pixels ({}%)'.format(
+        no_valid_pixels, no_total_pixels, valid_percent))
+
+    return 65535 * PIF_all.astype(numpy.uint16)
+
+
+def _PCA_fit_and_filter_single_band(candidate_band, reference_band,
+                                    valid_pixels_list, array_shape, lim):
+    ''' Performs PCA analysis, on the valid pixels and filters according
+    to the distance from the principle eigenvector, for a single band.
+    '''
+
+    cand_valid = candidate_band[valid_pixels_list]
+    ref_valid = reference_band[valid_pixels_list]
+
+    if logging.getLogger().getEffectiveLevel() <= logging.INFO:
+        logging.info('PCA Info: Original corrcoef = {}'.format(
+            numpy.corrcoef(cand_valid, ref_valid)[0, 1]))
+
+    fitted_pca = _PCA_fit_single_band(cand_valid, ref_valid)
+
+    if logging.getLogger().getEffectiveLevel() <= logging.INFO:
+        for component_no in xrange(len(fitted_pca.components_)):
+            logging.info('PCA Info: Component {} is {} with an explained '
+                         'variance of {}'.format(
+                            component_no,
+                            fitted_pca.components_[component_no],
+                            fitted_pca.explained_variance_ratio_[component_no])
+                         )
+
+    def split_into_batches(array, no_per_batch):
+        return [array[i:i + no_per_batch]
+                for i in xrange(0, len(array), no_per_batch)]
```

Comment:
This python loop can have performance consequence, especially since `no_per_batch = 10000`. If you can, try to incorporate this loop within the numpy api, so that the loop takes place in C. This also looks like the place where data should be read from disk, rather than keeping all in memory.